src/ralph/models/selector.py

```diff
@@ -0,0 +1,126 @@
+"""Model selector definition"""
+
+from collections import Counter
+from dataclasses import dataclass
+from importlib import import_module
+from itertools import chain
+from typing import Any, Union
+
+from pydantic import BaseModel
+
+from ralph.defaults import MODEL_PATH_SEPARATOR
+from ralph.utils import get_dict_value_from_path
+
+
+@dataclass(frozen=True)
+class LazyModelField:
+    """Represents a model field."""
+
+    path: tuple[str]
+
+    def __init__(self, path: str):
+        object.__setattr__(self, "path", tuple(path.split(MODEL_PATH_SEPARATOR)))
+
+
+@dataclass(frozen=True)
+class Rule:
+    """Represents a rule used for selection."""
+
+    field: LazyModelField
+    value: Union[LazyModelField, Any]
+
+    def check(self, event):
+        """Checks if event matches the rule.
+
+        Args:
+            event (dict): The event to check.
+        """
+
+        field = get_dict_value_from_path(event, self.field.path)
+        value = self.value
+        if isinstance(value, LazyModelField):
+            value = get_dict_value_from_path(event, value.path)
+        return field == value
+
+
+def selector(**fields_values):
+    """Returns a list of rules that should match in order to select an event.
+
+    Args:
+        fields_values (keyword arguments): the keywords represent the path to a
+            field and the values represent the value the field should match.
+    """
+
+    return [
+        Rule(LazyModelField(field), value) for field, value in fields_values.items()
+    ]
+
+
+class ModelSelector:
+    """Selects the matching model for a given event.
+
+    Attributes:
+        model_rules (dict): Stores the list of rules for each model.
+        decision_tree (dict): Stores the rule checking order for model selection.
+    """
+
+    model_rules: dict[BaseModel, list[Rule]] = {}
+
+    def __init__(self, module="ralph.models.edx"):
+        """Instantiates ModelSelector."""
+
+        for model in import_module(module).__dict__.values():
+            if isinstance(model, type) and model.__selector__:
```

Comment:
Can we be more specific?

```suggestion
            if isinstance(model, BaseModel) and model.__selector__:
```

Reply:
Good catch) thanks, yes indeed, have added `issubclass(model, BaseModel)`, maybe I should rename model to `model_class` ?

Reply:
Indeed the typing was misleading, model is of class type, updated the typing to Type[BaseModel] and replaced model with model_class name.